zerver/tests/test_subs.py

```diff
@@ -3235,93 +3091,128 @@ def test_subscriptions_add_stream_with_null(self) -> None:
             result, f"Stream name '{stream_name}' contains NULL (0x00) characters."
         )
 
-    def test_user_settings_for_adding_streams(self) -> None:
+    def _test_user_settings_for_adding_streams(self, stream_policy: str, invite_only: bool) -> None:
+        method = "can_create_private_streams" if invite_only else "can_create_public_streams"
         do_set_realm_property(
-            self.test_user.realm, "create_stream_policy", Realm.POLICY_ADMINS_ONLY, acting_user=None
+            self.test_user.realm, stream_policy, Realm.POLICY_ADMINS_ONLY, acting_user=None
         )
-        with mock.patch("zerver.models.UserProfile.can_create_streams", return_value=False):
-            result = self.common_subscribe_to_streams(self.test_user, ["stream1"], allow_fail=True)
+
+        with mock.patch(f"zerver.models.UserProfile.{method}", return_value=False):
+            result = self.common_subscribe_to_streams(
+                self.test_user, ["stream1"], invite_only=invite_only, allow_fail=True
+            )
             self.assert_json_error(result, "Insufficient permission")
 
-        with mock.patch("zerver.models.UserProfile.can_create_streams", return_value=True):
-            self.common_subscribe_to_streams(self.test_user, ["stream2"])
+        with mock.patch(f"zerver.models.UserProfile.{method}", return_value=True):
+            self.common_subscribe_to_streams(self.test_user, ["stream2"], invite_only=invite_only)
 
         # User should still be able to subscribe to an existing stream
-        with mock.patch("zerver.models.UserProfile.can_create_streams", return_value=False):
-            self.common_subscribe_to_streams(self.test_user, ["stream2"])
+        with mock.patch(f"zerver.models.UserProfile.{method}", return_value=False):
+            self.common_subscribe_to_streams(self.test_user, ["stream2"], invite_only=invite_only)
 
-    def test_user_settings_for_creating_streams(self) -> None:
+    def test_user_settings_for_adding_streams(self) -> None:
```

Comment:
I would recommend you split this set of tests into two. What happens when you club tests like this is when the first one fails the second one won't even run and ideally you would want it to run since its a different test.

I see at other places too you have clubbed tests like this. I would recommend fixing those two :)

Reply:
I don't think this is important one way or another; might be a good discussion on "automated testing".

(The question is whether we should wrap the two calls to the `_` method in different test functions).